File: android/app/src/main/java/com/certisgroup/cathy/reactutils/modules/crypto/RsaModule.java
/*
 * Copyright (c) 2020 Certis CISCO Security Pte Ltd
 * All rights reserved.
 *
 * This software is the confidential and proprietary information of
 * Certis CISCO Security Pte Ltd. ("Confidential Information").
 * You shall not disclose such Confidential Information and shall use
 * it only in accordance with the terms of the license agreement you
 * entered into with Certis CISCO Security Pte Ltd.
 */
package com.certisgroup.cathy.reactutils.modules.crypto;

import android.util.Base64;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;

import java.math.BigInteger;
import java.security.GeneralSecurityException;
import java.security.Key;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.AlgorithmParameterSpec;
import java.security.spec.KeySpec;
import java.security.spec.MGF1ParameterSpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.RSAKeyGenParameterSpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import javax.crypto.Cipher;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;

/**
 * RN module for RSA algorithms
 *
 * @author Lingqi
 */
public class RsaModule extends ReactContextBaseJavaModule {

    private static final String PKCS1_v1_5 = "PKCS1-v1_5";
    private static final Map<String, String> SIG_DIGEST_MAP = sigDigestMap();
    private static Map<String, String> sigDigestMap() {
        Map<String, String> map = new HashMap<>();
        map.put("SHA-1", "SHA1withRSA");
        map.put("SHA-256", "SHA256withRSA");
        map.put("SHA-384", "SHA384withRSA");
        map.put("SHA-512", "SHA512withRSA");
        return map;
    }
    private static final Map<String, MGF1ParameterSpec> MGF1_SPEC_MAP = mgf1SpecMap();
    private static Map<String, MGF1ParameterSpec> mgf1SpecMap() {
        Map<String, MGF1ParameterSpec> map = new HashMap<>();
        map.put("SHA-1", MGF1ParameterSpec.SHA1);
        map.put("SHA-256", MGF1ParameterSpec.SHA256);
        map.put("SHA-384", MGF1ParameterSpec.SHA384);
        map.put("SHA-512", MGF1ParameterSpec.SHA512);
        return map;
    }

    private final Map<String, PrivateKey> privateKeyCache;
    private final Map<String, PublicKey> publicKeyCache;

    public RsaModule(@NonNull ReactApplicationContext reactContext) {
        super(reactContext);
        privateKeyCache = new HashMap<>();
        publicKeyCache = new HashMap<>();
    }

    @NonNull
    @Override
    public String getName() {
        return "CryptoRSA";
    }

    @ReactMethod
    public void sign(ReadableMap params, String uuid, String data, Promise promise) {
        if (!PKCS1_v1_5.equals(params.getString("scheme"))) {
            promise.reject("RSA_SIGN_ERROR", "Scheme not supported");
            return;
        }
        try {
            String hash = params.getString("hash");
            String algorithm = SIG_DIGEST_MAP.get(hash);
            PrivateKey privateKey = privateKeyCache.get(uuid);
            if (algorithm == null || privateKey == null) {
                throw new IllegalArgumentException("Hash or Key not valid");
            }
            byte[] dataBytes = Base64.decode(data, Base64.DEFAULT);
            Signature sig = Signature.getInstance(algorithm);
            sig.initSign(privateKey);
            sig.update(dataBytes);
            byte[] sigBytes = sig.sign();
            String signature = Base64.encodeToString(sigBytes, Base64.NO_WRAP);
            promise.resolve(signature);
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("RSA_SIGN_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void verify(ReadableMap params, String uuid, String data, String signature,
                       Promise promise) {
        if (!PKCS1_v1_5.equals(params.getString("scheme"))) {
            promise.reject("RSA_VERIFY_ERROR", "Scheme not supported");
            return;
        }
        try {
            String hash = params.getString("hash");
            String algorithm = SIG_DIGEST_MAP.get(hash);
            PublicKey publicKey = publicKeyCache.get(uuid);
            if (algorithm == null || publicKey == null) {
                throw new IllegalArgumentException("Hash or Key not valid");
            }
            byte[] dataBytes = Base64.decode(data, Base64.DEFAULT);
            byte[] sigBytes = Base64.decode(signature, Base64.DEFAULT);
            Signature sig = Signature.getInstance(algorithm);
            sig.initVerify(publicKey);
            sig.update(dataBytes);
            boolean valid = sig.verify(sigBytes);
            promise.resolve(valid);
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("RSA_VERIFY_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void encrypt(ReadableMap params, String uuid, String data, Promise promise) {
        try {
            String padding = params.getString("padding");
            PublicKey publicKey = publicKeyCache.get(uuid);
            if (padding == null || publicKey == null) {
                throw new IllegalArgumentException("Padding or Key not valid");
            }
            String result;
            switch (padding) {
                case "PKCS1":
                    result = this.pkcs1Cipher(Cipher.ENCRYPT_MODE, publicKey, data);
                    break;
                case "OAEP":
                    String hash = params.getString("hash");
                    result = this.oaepCipher(Cipher.ENCRYPT_MODE, hash, publicKey, data);
                    break;
                default:
                    promise.reject("RSA_ENCRYPT_ERROR", "Padding not supported");
                    return;
            }
            promise.resolve((result));
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("RSA_ENCRYPT_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void decrypt(ReadableMap params, String uuid, String data, Promise promise) {
        try {
            String padding = params.getString("padding");
            PrivateKey privateKey = privateKeyCache.get(uuid);
            if (padding == null || privateKey == null) {
                throw new IllegalArgumentException("Padding or Key not valid");
            }
            String result;
            switch (padding) {
                case "PKCS1":
                    result = this.pkcs1Cipher(Cipher.DECRYPT_MODE, privateKey, data);
                    break;
                case "OAEP":
                    String hash = params.getString("hash");
                    result = this.oaepCipher(Cipher.DECRYPT_MODE, hash, privateKey, data);
                    break;
                default:
                    promise.reject("RSA_DECRYPT_ERROR", "Padding not supported");
                    return;
            }
            promise.resolve((result));
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("RSA_DECRYPT_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void generateKeyPair(int modulusSize, Promise promise) {
        try {
            BigInteger publicExponent = RSAKeyGenParameterSpec.F4;
            AlgorithmParameterSpec rsaSpec = new RSAKeyGenParameterSpec(modulusSize, publicExponent);
            KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance("RSA");
            keyPairGen.initialize(rsaSpec);
            KeyPair keyPair = keyPairGen.genKeyPair();
            String privateId = UUID.randomUUID().toString();
            String publicId = UUID.randomUUID().toString();
            privateKeyCache.put(privateId, keyPair.getPrivate());
            publicKeyCache.put(publicId, keyPair.getPublic());
            WritableMap privateKey = Arguments.createMap();
            privateKey.putString("uuid", privateId);
            privateKey.putString("format", "pkcs8");
            WritableMap publicKey = Arguments.createMap();
            publicKey.putString("uuid", publicId);
            publicKey.putString("format", "spki");
            WritableMap map = Arguments.createMap();
            map.putMap("privateKey", privateKey);
            map.putMap("publicKey", publicKey);
            promise.resolve(map);
        } catch (GeneralSecurityException e) {
            promise.reject("RSA_GENERATE_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void importKey(String format, String key, Promise promise) {
        try {
            byte[] keyBytes = Base64.decode(key, Base64.DEFAULT);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            String uuid = UUID.randomUUID().toString();
            switch (format) {
                case "pkcs8":
                    KeySpec pkcs8KeySpec = new PKCS8EncodedKeySpec(keyBytes);
                    PrivateKey privateKey = keyFactory.generatePrivate(pkcs8KeySpec);
                    privateKeyCache.put(uuid, privateKey);
                    break;
                case "spki":
                    KeySpec x509KeySpec = new X509EncodedKeySpec(keyBytes);
                    PublicKey publicKey = keyFactory.generatePublic(x509KeySpec);
                    publicKeyCache.put(uuid, publicKey);
                    break;
                default:
                    promise.reject("RSA_IMPORT_ERROR", "Format not supported");
                    return;
            }
            WritableMap map = Arguments.createMap();
            map.putString("uuid", uuid);
            map.putString("format", format);
            promise.resolve(map);
        } catch (GeneralSecurityException | IllegalArgumentException e) {
            promise.reject("RSA_IMPORT_ERROR", e.getMessage(), e);
        }
    }

    @ReactMethod
    public void exportKey(String format, String uuid, Promise promise) {
        Key key;
        switch (format) {
            case "pkcs8":
                key = privateKeyCache.get(uuid);
                break;
            case "spki":
                key = publicKeyCache.get(uuid);
                break;
            default:
                promise.reject("RSA_EXPORT_ERROR", "Format not supported");
                return;
        }
        if (key != null) {
            byte[] keyBytes = key.getEncoded();
            String keyString = Base64.encodeToString(keyBytes, Base64.NO_WRAP);
            promise.resolve(keyString);
        } else {
            promise.reject("RSA_EXPORT_ERROR", "Key does not exist");
        }
    }

    private String pkcs1Cipher(int op, Key key, String data)
            throws GeneralSecurityException, IllegalArgumentException {
        Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        cipher.init(op, key);
        byte[] input = Base64.decode(data, Base64.DEFAULT);
        byte[] result = cipher.doFinal(input);
        return Base64.encodeToString(result, Base64.NO_WRAP);
    }

    private String oaepCipher(int op, String hash, Key key, String data)
            throws GeneralSecurityException, IllegalArgumentException {
        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPPadding");
        MGF1ParameterSpec mgf1Spec = MGF1_SPEC_MAP.get(hash);
        if (mgf1Spec == null) {
            throw new IllegalArgumentException("Hash not valid");
        }
        OAEPParameterSpec oaepSpec = new OAEPParameterSpec(hash, "MGF1", mgf1Spec,
            PSource.PSpecified.DEFAULT);
        cipher.init(op, key, oaepSpec);
        byte[] input = Base64.decode(data, Base64.DEFAULT);
        byte[] result = cipher.doFinal(input);
        return Base64.encodeToString(result, Base64.NO_WRAP);
    }
}

File: ios/cathy/RCTUtils/Modules/Crypto/RsaModule.swift
/*
 * Copyright (c) 2020 Certis CISCO Security Pte Ltd
 * All rights reserved.
 *
 * This software is the confidential and proprietary information of
 * Certis CISCO Security Pte Ltd. ("Confidential Information").
 * You shall not disclose such Confidential Information and shall use
 * it only in accordance with the terms of the license agreement you
 * entered into with Certis CISCO Security Pte Ltd.
 */
import Foundation
import Security
import CommonCrypto

/**
 * RN module for RSA algorithms
 *
 * @author Lingqi
 */
@objc(CryptoRSA)
class RsaModule: NSObject {
    
    private let PKCS1_v1_5 = "PKCS1-v1_5";
    private let SIG_HASH_DICT: [String: SecPadding] = [
        "SHA-1": .PKCS1SHA1,
        "SHA-256": .PKCS1SHA256,
        "SHA-384": .PKCS1SHA384,
        "SHA-512": .PKCS1SHA512
    ]
    private let DIGEST_LEN_DICT = [
        "SHA-1": CC_SHA1_DIGEST_LENGTH,
        "SHA-256": CC_SHA256_DIGEST_LENGTH,
        "SHA-384": CC_SHA384_DIGEST_LENGTH,
        "SHA-512": CC_SHA512_DIGEST_LENGTH
    ];
    private let SHA_FUNC_DICT = [
        "SHA-1": CC_SHA1,
        "SHA-256": CC_SHA256,
        "SHA-384": CC_SHA384,
        "SHA-512": CC_SHA512
    ]
    private let PRIVATE_TAG = Data("PRIVATE_KEY".utf8)
    private let PUBLIC_TAG = Data("PUBLIC_KEY".utf8)
    
    private var privateKeyCache: [String: SecKey]
    private var publicKeyCache: [String: SecKey]
    
    override init() {
        privateKeyCache = [:]
        publicKeyCache = [:]
        super.init()
    }
    
    @objc(sign:::::)
    func sign(params: [String: String], uuid: String, data: String,
              resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if params["scheme"] != PKCS1_v1_5 {
            reject("RSA_SIGN_ERROR", "Scheme not supported", nil)
            return
        }
        do {
            guard let hash = params["hash"],
                let padding = SIG_HASH_DICT[hash],
                let key = privateKeyCache[uuid],
                let dataBuffer = Data(base64Encoded: data) else {
                    throw SecError.errParam
            }
            let dataBytes = [UInt8](dataBuffer)
            var digest = [UInt8](repeating: 0, count: Int(DIGEST_LEN_DICT[hash]!))
            _ = SHA_FUNC_DICT[hash]!(dataBytes, CC_LONG(dataBytes.count), &digest)
            var sigLen = SecKeyGetBlockSize(key)
            var sig = [UInt8](repeating: 0, count: sigLen)
            let status = SecKeyRawSign(key, padding, digest, digest.count, &sig, &sigLen)
            if status != errSecSuccess {
                throw SecError(rawValue: status)!
            }
            let sigData = Data(bytes: sig, count: sigLen)
            let signature = sigData.base64EncodedString()
            resolve(signature)
        } catch let error {
            reject("RSA_SIGN_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(verify::::::)
    func verify(params: [String: String], uuid: String, data: String, signature: String,
                resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        if params["scheme"] != PKCS1_v1_5 {
            reject("RSA_VERIFY_ERROR", "Scheme not supported", nil)
            return
        }
        do {
            guard let hash = params["hash"],
                let padding = SIG_HASH_DICT[hash],
                let key = publicKeyCache[uuid],
                let dataBuffer = Data(base64Encoded: data),
                let sigData = Data(base64Encoded: signature) else {
                    throw SecError.errParam
            }
            let dataBytes = [UInt8](dataBuffer)
            var digest = [UInt8](repeating: 0, count: Int(DIGEST_LEN_DICT[hash]!))
            _ = SHA_FUNC_DICT[hash]!(dataBytes, CC_LONG(dataBytes.count), &digest)
            let sig = [UInt8](sigData)
            let status = SecKeyRawVerify(key, padding, digest, digest.count, sig, sig.count)
            if status != errSecSuccess && status != errSSLCrypto {
                throw SecError(rawValue: status)!
            }
            let verified = status == errSecSuccess
            resolve(verified)
        } catch let error {
            reject("RSA_VERIFY_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(encrypt:::::)
    func encrypt(params: [String: String], uuid: String, data: String,
                 resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            guard let padding = params["padding"],
                let key = publicKeyCache[uuid] else {
                    throw SecError.errParam
            }
            let result: String
            switch padding {
                case "PKCS1":
                    result = try cipher(op: .encrypt, padding: .PKCS1, key: key, data: data)
                case "OAEP":
                    if params["hash"] != "SHA-1" { throw SecError.errParam }
                    result = try cipher(op: .encrypt, padding: .OAEP, key: key, data: data)
                default:
                    reject("RSA_ENCRYPT_ERROR", "Padding not supported", nil)
                    return
            }
            resolve(result)
        } catch let error {
            reject("RSA_ENCRYPT_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(decrypt:::::)
    func decrypt(params: [String: String], uuid: String, data: String,
                 resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            guard let padding = params["padding"],
                let key = privateKeyCache[uuid] else {
                    throw SecError.errParam
            }
            let result: String
            switch padding {
                case "PKCS1":
                    result = try cipher(op: .decrypt, padding: .PKCS1, key: key, data: data)
                case "OAEP":
                    if params["hash"] != "SHA-1" { throw SecError.errParam }
                    result = try cipher(op: .decrypt, padding: .OAEP, key: key, data: data)
                default:
                    reject("RSA_DECRYPT_ERROR", "Padding not supported", nil)
                    return
            }
            resolve(result)
        } catch let error {
            reject("RSA_DECRYPT_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(generateKeyPair:::)
    func generateKeyPair(modulusSize: Int, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            let parameters: [CFString: Any] = [
                kSecAttrKeyType: kSecAttrKeyTypeRSA,
                kSecAttrKeySizeInBits: modulusSize
            ]
            var publicKey: SecKey?
            var privateKey: SecKey?
            let status = SecKeyGeneratePair(parameters as CFDictionary, &publicKey, &privateKey)
            if status != errSecSuccess {
                throw SecError(rawValue: status)!
            }
            let publicId = UUID().uuidString
            let privateId = UUID().uuidString
            publicKeyCache[publicId] = publicKey!
            privateKeyCache[privateId] = privateKey!
            let dict = [
                "privateKey": [
                    "uuid": privateId,
                    "format": "pkcs8"
                ],
                "publicKey": [
                    "uuid": publicId,
                    "format": "spki"
                ]
            ]
            resolve(dict)
        } catch let error {
            reject("RSA_GENERATE_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(importKey::::)
    func importKey(format: String, key: String,
                   resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        do {
            guard let keyData = Data(base64Encoded: key) else {
                throw SecError.errParam
            }
            let uuid = UUID().uuidString
            switch format {
                case "pkcs8":
                    let keyClass = kSecAttrKeyClassPrivate
                    let key = try self.createKeyWithData(keyData: keyData, keyClass: keyClass)
                    privateKeyCache[uuid] = key
                case "spki":
                    let keyClass = kSecAttrKeyClassPublic
                    let key = try self.createKeyWithData(keyData: keyData, keyClass: keyClass)
                    publicKeyCache[uuid] = key
                default:
                    reject("RSA_IMPORT_ERROR", "Format not supported", nil)
                    return
            }
            let dict = [
                "uuid": uuid,
                "format": format
            ]
            resolve(dict)
        } catch let error {
            reject("RSA_IMPORT_ERROR", error.localizedDescription, error)
        }
    }
    
    @objc(exportKey::::)
    func exportKey(format: String, uuid: String,
                   resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        let key: SecKey?
        let keyClass: CFString
        switch format {
            case "pkcs8":
                key = privateKeyCache[uuid]
                keyClass = kSecAttrKeyClassPrivate
            case "spki":
                key = publicKeyCache[uuid]
                keyClass = kSecAttrKeyClassPublic
            default:
                reject("RSA_EXPORT_ERROR", "Format not supported", nil)
                return
        }
        if let key = key {
            let keyData = self.copyKeyToExternal(key: key, keyClass: keyClass)
            let keyString = keyData.base64EncodedString()
            resolve(keyString)
        } else {
            reject("RSA_EXPORT_ERROR", "Key does not exist", nil)
        }
    }
    
    private func cipher(op: SecOperation, padding: SecPadding, key: SecKey, data: String) throws -> String {
        guard let dataBuffer = Data(base64Encoded: data) else {
            throw SecError.errParam
        }
        let input = [UInt8](dataBuffer)
        var textLen = SecKeyGetBlockSize(key)
        var text = [UInt8](repeating: 0, count: textLen)
        let status: OSStatus
        switch op {
            case .encrypt:
                status = SecKeyEncrypt(key, padding, input, input.count, &text, &textLen)
            case .decrypt:
                status = SecKeyDecrypt(key, padding, input, input.count, &text, &textLen)
        }
        if status != errSecSuccess {
            throw SecError(rawValue: status)!
        }
        let result = Data(bytes: text, count: textLen)
        return result.base64EncodedString()
    }
    
    /// Don't use SecKeyCreateWithData, it can't handle pkcs8
    private func createKeyWithData(keyData: Data, keyClass: CFString) throws -> SecKey {
        guard let keyData = keyClass == kSecAttrKeyClassPrivate ?
            Pkcs8Util.stripPkcs8Header(keyData: keyData) :
            Pkcs8Util.stripSpkiHeader(keyData: keyData) else {
                throw SecError.errDecode
        }
        let tag = keyClass == kSecAttrKeyClassPrivate ? PRIVATE_TAG : PUBLIC_TAG
        self.deleteSecKey(tag: tag)
        let attributes: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrApplicationTag: tag,
            kSecAttrKeyClass: keyClass,
            kSecValueData: keyData,
            kSecReturnRef: true
        ]
        var key: CFTypeRef?
        let status = SecItemAdd(attributes as CFDictionary, &key)
        if status != errSecSuccess {
            throw SecError(rawValue: status)!
        }
        self.deleteSecKey(tag: tag)
        return key as! SecKey
    }
    
    /// Don't use SecKeyCopyExternalRepresentation, it can't handle pkcs8
    private func copyKeyToExternal(key: SecKey, keyClass: CFString) -> Data {
        let tag = keyClass == kSecAttrKeyClassPrivate ? PRIVATE_TAG : PUBLIC_TAG
        self.deleteSecKey(tag: tag)
        let attributes: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrApplicationTag: tag,
            kSecAttrKeyClass: keyClass,
            kSecValueRef: key,
            kSecReturnData: true
        ]
        var keyData: CFTypeRef?
        SecItemAdd(attributes as CFDictionary, &keyData)
        self.deleteSecKey(tag: tag)
        if keyClass == kSecAttrKeyClassPrivate {
            return Pkcs8Util.addPkcs8Header(keyData: keyData as! Data)
        } else {
            return Pkcs8Util.addSpkiHeader(keyData: keyData as! Data)
        }
    }
    
    private func deleteSecKey(tag: Data) {
        let query: [CFString: Any] = [
            kSecClass: kSecClassKey,
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrApplicationTag: tag
        ]
        SecItemDelete(query as CFDictionary)
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}

private enum SecOperation {
    case encrypt
    case decrypt
}

private enum SecError: OSStatus, Error {
    case errUnimplemented = -4
    case errIO            = -36
    case errParam         = -50
    case errAllocate      = -108
    case errBadReq        = -909
    case errInternalComponent = -2070
    case errCrypto        = -9809
    case errNotAvailable  = -25291
    case errDuplicateItem = -25299
    case errItemNotFound  = -25300
    case errDecode        = -26275
}

private let ERROR_DICT: [SecError: String] = [
    .errUnimplemented: "Function or operation not implemented",
    .errIO: "I/O error (bummers)",
    .errParam: "One or more parameters passed to a function were not valid",
    .errAllocate: "Failed to allocate memory",
    .errBadReq: "Bad parameter or invalid state for operation",
    .errInternalComponent: "An internal component experienced an error",
    .errCrypto: "Underlying cryptographic error",
    .errNotAvailable: "No keychain is available",
    .errDuplicateItem: "The specified item already exists in the keychain",
    .errItemNotFound: "The specified item could not be found in the keychain",
    .errDecode: "Unable to decode the provided data",
]

extension SecError: LocalizedError {
    var errorDescription: String? {
        return ERROR_DICT[self]
    }
}
